### ISSUE TITLE
EmbeddedId in findByIdIn does not seem to be working

### ISSUE THREAD

**Issue by berkum** [micronaut-projects/micronaut-data#594]
When trying to use findByIdIn with an EmbeddedId, micronaut does not seem to understand it is dealing with an embedded id and does not generate the correct SQL.

### Task List
- [ ] Steps to reproduce provided
I used the micronaut source code to reproduce.  There is one entity with an EmbeddableId, class Project.
I added the method signature to io.micronaut.data.tck.repositories.ProjectRepository :
 List<Project> findByIdIn(List<ProjectId> ids);

Then in H2CompositePrimaryKeySpec, I used the first test ("test CRUD with composite ID")
and inserted my call  : 
 project = projectRepository.findById(id).orElse(null)  // already in place from micronaut
 projectRepository.findByIdIn(Arrays.asList(id));  // This is my call

### Expected Behaviour

It should have found my record but most importantly, it should not generate an error.

### Actual Behaviour
SQL Error preparing Query: Column "PROJECT_.PROJECT_ID" not found; SQL statement:

If you look at the SQL generated for findById, the where clause is OK : 
`[Test worker] DEBUG io.micronaut.data.query - Executing Query: SELECT project_.`project_id_department_id`,project_.`project_id_project_id`,project_.`name`,UPPER(org) AS org FROM `project` project_ WHERE (project_.project_id_department_id = ? AND project_.project_id_project_id = ?)`

And then the generated SQL for findByIdIn (incorrect) :
`Executing Query: SELECT project_.`project_id_department_id`,project_.`project_id_project_id`,project_.`name`,UPPER(org) AS org FROM `project` project_ WHERE (project_.project_id IN(?))` 

### Environment Information

- **Operating System**: macOS Catalina
- **Micronaut Version:** micronaut-data 1.0.2
- **JDK Version:** openjdk version "1.8.0_242"

**Comment by jaksah** [comment]
I'm experiencing the same issue, using micronaut-data 1.1.3

**Comment by jaksah** [comment]
We've managed to do a workaround using `JdbcOperations`. Had to add some "reactive fluff" as well to since we're using the `RxJavaCrudRepository`. Something in line with this:
```
public abstract class EntityCrudRepository implements RxJavaCrudRepository<Entity, EntityKey> {
    private final JdbcOperations jdbcOperations;
    private final ExecutorService executorService;

    protected EntityCrudRepository(
        final JdbcOperations jdbcOperations,
        @Named(TaskExecutors.IO) final ExecutorService executorService
    ) {
        this.jdbcOperations = jdbcOperations;
        this.executorService = executorService;
    }

    @Transactional
    public Observable<Entity> findByKeyUserIdIn(
        final List<String> userIds
    ) {
        return Observable.<Entity>create(emitter -> {
                try {
                    jdbcOperations.prepareStatement(
                        "SELECT * FROM entity WHERE key_user_id = ANY (?)",
                        statement -> {
                            Array array = statement.getConnection()
                                              .createArrayOf(
                                                  "varchar",
                                                  userIds.toArray(new String[0])
                                              );
                            statement.setArray(1, array);
                            return jdbcOperations.entityStream(
                                statement.executeQuery(),
                                Entity.class
                            );
                        }
                    ).forEach(emitter::onNext);
                    emitter.onComplete();
                } catch (DataAccessException e) {
                    emitter.onError(e);
                }
            }
        ).subscribeOn(Schedulers.from(executorService));
    }
}
```

**Comment by VishGov** [comment]
Hi All, 

I am using Micronaut 4.2.1 with micronaut data and micronaut jdbc. 

If I have a repository like this with CompositeKey being an EmbeddedId

@JdbcRepository(dialect = Dialect.POSTGRES)
public interface TestRepository extends CrudRepository<TestEntity, CompositeKey> {

    List<TestEntity> findByTestCompositeKeyIn(List<CompositeKey> testCompositeKeys);
}

When I am trying findByXIn queries, the List<?> is not being taken in the query and I am getting the following error

org.postgresql.util.PSQLException: Can't infer the SQL type to use for an instance of com.entity.data.CompositeKey. Use setObject() with an explicit Types value to specify the type to use.

The error maybe different, but the fundamental issue of the List not being considered properly to create the correct WHERE IN query is still present. Hope a fix for this issue will be able to resolve the error I have mentioned as well.

**Comment by VishGov** [comment]
@radovanradic I see you have mentioned this ticket as part of a fix. Do you know when this is planned to be released?

**Comment by radovanradic** [comment]
No, had PR but with incorrect approach so not sure when the fix can be done.

**Comment by VishGov** [comment]
@radovanradic I was reading the PR and you mentioned the following

> "Don't think it ever worked, this issue is from 2020.
> Tried approach 2 (that is how Hibernate SQL looks like in the end) and that was not doable with how parameter expanding is implemented."

Do you mean in micronaut code currently if a list of composite key objects are given
[ {field1:B, field2:1}, {field1:C, field2:2} ]

it cannot be expanded and the below SQL cannot be created?
SELECT *
FROM table1
WHERE (table1.field1, table1.field2) IN ( ('B',1), ('C',2) );

**Comment by radovanradic** [comment]
Yes, and it requires more changes to support it. Not sure also that all DBs support this IN clause with multiple columns.

**Comment by VishGov** [comment]
I believe Spring Data is already handling this case, so it would be great Micronaut Data can also support this soon.

Maybe they are expanding and handling it like this to make it generic? (Not sure)
SELECT  * 
FROM    tbl1
WHERE   ID1 = 1 AND ID2 = 1 OR
        ID1 = 1 AND ID2 = 2 OR
        ID1 = 3 AND ID2 = 5 OR
        ID1 = 2 AND ID2 = 2

**Comment by gs-kug** [comment]
I have this problem but with inserts:

```
Executing SQL query: INSERT INTO "inventory_movement_product" ("identity","created_date","last_modified_date","created_by","quantity","product_id","inventory_movement_id") VALUES (?,?,?,?,?,?,?)
16:01:07.086 [default-nioEventLoopGroup-1-2] TRACE io.micronaut.data.query - Binding parameter at position 1 to value InventoryMovementID(productId=019437ca-6383-756d-875e-681a2dcea205, inventoryMovementId=019437cb-8758-746f-a052-673d40f3b53b) with data type: OBJECT
16:01:07.098 [default-nioEventLoopGroup-1-2] ERROR c.k.m.p.api.AppExceptionHandler - Error executing PERSIST: Error executing PERSIST: Unable to set PreparedStatement value: Can't infer the SQL type to use for an instance of com.kugelbit.mercado.infrastructure.database.domain.InventoryMovementProductID. Use setObject() with an explicit Types value to specify the type to use.
io.micronaut.data.exceptions.DataAccessException: Error executing PERSIST: Error executing PERSIST: Unable to set PreparedStatement value: Can't infer the SQL type to use for an instance of com.kugelbit.mercado.infrastructure.database.domain.InventoryMovementProductID. Use setObject() with an explicit Types value to specify the type to use.
	at io.micronaut.data.runtime.operations.internal.BaseOperations.failed(BaseOperations.java:141)
```

ID is embedded. I will try a workaround by modelling id differently, this is a many to many relationship so makes sense to have composite primary keys. Please take special care of this issue, as this prevents some common models of relationships.

**Comment by dstepanov** [comment]
Design notes for whoever picks this up, after tracing the current root cause against the 5.1.x branch.

**Root cause, precisely:**
- `AbstractSqlLikeQueryBuilder.visitIn` (data-model/src/main/java/io/micronaut/data/model/query/builder/sql/AbstractSqlLikeQueryBuilder.java, ~line 2706) appends a single property reference regardless of whether it resolves to a multi-column embedded/composite property, producing `col1 IN (?)` instead of matching all key columns.
- `DefaultSqlPreparedQuery.prepare()` (data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java, ~line 209) is where the runtime expansion happens — for an expandable parameter it currently only knows how to repeat a single `?` placeholder per collection item; it has no notion of a per-item "arity" (K placeholders instead of 1).

**Why the previously attempted approach didn't land:** a dialect-specific tuple-IN syntax (`(col1,col2) IN ((?,?),(?,?))`) isn't universally supported and, per the PR discussion here, wasn't doable with how parameter expansion is currently implemented without deeper changes.

**Suggested design instead — OR-of-ANDs:**
```sql
WHERE (col1 = ? AND col2 = ?) OR (col1 = ? AND col2 = ?) OR ...
```
This is portable to every SQL dialect Micronaut Data supports, unlike tuple-IN. It needs:
1. `visitIn` to detect a multi-column property path and emit a repeatable AND-group marker instead of a plain `IN`.
2. The expansion arity threaded through to `DefaultSqlPreparedQuery.prepare()` so it repeats K placeholders per item (grouped and OR-joined) instead of 1 comma-separated placeholder per item.
3. The binding side to reuse the existing composite-key equality-binding logic already used for `findById(compositeKey)` (which already knows how to bind K sub-values from one embedded object), rather than the current single-value-per-item IN logic.

**Scoping suggestion:** start with `deleteAll(Collection<T>)` only (the exact original report here, one call site) before touching `findByIdIn`/general derived-query IN clauses, given how central this code path is and how easy it'd be to introduce a subtle SQL-correctness regression across dialects.

Not attempting an implementation in this pass — leaving this as a scoped starting point given the history of attempts here.
